README format from md to rst

fragmuffin opened the pull request:
addresses issue #287

fragmuffin commented:
oh! I just noticed that PyPI **can** handle Markdown :roll_eyes: 
Is that new?; I thought it only handled rst.
Not sure if this is a good change then... perhaps I should have stuck with md.

westurner commented:
Sphinx can also handle markdown; though roles and directives are not implemented for markdown, so it makes sense for README to be in RST so that it can easily be included in the docs verbatim.

Thanks!

dcowden commented:
Ok so the consensus is that this is ready to merge?

fragmuffin commented:
@dcowden 

I think it is :+1:

There are also minor changes to:
- Dockerfile
- setup.py

But if there are any unforeseen issues with those, I'm sure they can be fixed easily in the next release.

fragmuffin commented:
Hmph :disappointed:. I've played around a bit with the images, but I've come around full circle.
It seems the attributes are being picked up when rendering
eg: see `alt="Binder"` for pill in [I just want to try things out](https://github.com/fragmuffin/cadquery/tree/feature/readme-rst#i-just-want-to-try-things-out)
But `:align:` and `:width:` are (mostly) ignored.

*mostly?
Both `:width:` and `:align:` attributes _are_ rendered in the logo `img` tag at the top, but the `align` attribute is deprecated in HTML5 (see: github/markup#163).
The question remains: how to center align an image when rendered to HTML5?
However, further down in the readme, _neither_ `:width:` nor `:align:` are rendered to the KiCad `img` tag.

I'm so confused :confused: 

Does anyone have:
- experience with rendering images in reStructured text?, or
- a better idea for a solution to #287

dcowden commented:
@fragmuffin the cadquery spinx docs render images. For example, quickstart:
https://github.com/dcowden/cadquery/blob/master/doc/quickstart.rst

For a more advanced case, the [cq examples](https://github.com/dcowden/cadquery/blob/master/doc/examples.rst) uses a custom directive ([cqplot](https://github.com/dcowden/cadquery/blob/master/cadquery/cq_directive.py)) that executes cq script and then creates an image and embeds it.

fragmuffin commented:
OK!, I'm so very done with this...
I gave up with `image` and `figure` and reverted any scaled images to use the `.. raw:: html` directive.

So the equivalent of:

```rst
.. image:: https://foo.com/bar.png

    :width: 100px
    :align: center
```

which doesn't render properly, ref:  github/markup#295

```html
.. raw:: html

    <div align="center">
        <img src="https://foo.com/bar.png" width="100px" />
    </div>
```

Now... At this point I'm assuming this will render the same in PyPI as it does in github.
All link targets and image sources are absolute (same as before), so there shouldn't be any problems there.

dcowden commented:
+1 from me, thanks for the work @fragmuffin !
@jmwright any objections merging this?

jmwright commented:
@dcowden I'm not sure I fully understand all the implications of this change, but I'm ok with it if you are.

fragmuffin commented:
@jmwright 

Ultimately there is zero change to an end-user's experience of `cadquery`, with the exception of PyPI's rendering of the README (see #287)

PyPI more directly supports _rst_ over _md_, hence the change.
Everything else discussed in this PR is specifically to do with [GitHub's neglect of _rst_](https://github.com/github/markup/issues/295), which is unfortunate, but manageable.

In hindsight, I'm not sure this PR was worth the effort, but it _should_ at least make the readme render the same on GitHub & PyPI.
\+ I can take what I've learned and apply it to my other projects 😉 

I just hope the `.. raw:: html` directive renders properly on PyPI.
If it doesn't 😠 , I'm going to make a dummy module on the PyPI sandbox server **just** to try different rst stuff.

westurner commented:
https://github.com/pypa/readme_renderer/blob/master/README.rst#check-description-locally

```bash
$ pip install readme_renderer
$ python setup.py check -r -s
```

AFAIU, raw HTML probably doesn't work on PyPI (so that JS couldn't e.g log you out).

Image rendering on the GitHub RST renderer (docutils, too, IIRC) view is less important than PyPI compatibility, IMHO.

westurner commented:
There's a test.pypi.org server that doesn't affect the real index; test pypi packages don't get mirrored many times all over the world).
https://packaging.python.org/guides/using-testpypi/

fragmuffin commented:
@westurner 

```
$ python setup.py -r -s
running check
warning: Check: The project's long_description has invalid markup which will not be rendered on PyPI. The following syntax errors were detected:
line 1: Warning: "raw" directive disabled.


error: Please correct your package.
```

oh joy :cry:

the journey continues... this is a good test though! I'll add it to travis

fragmuffin commented:
Hi all, I thought this was sorted months ago!
But I totally left it hanging... my bad... so I'm finishing this asap.

The issue is simple:

> reStructured text [when it comes to image properties] has very limited support by PyPI

So simply scaling an image isn't possible (if I'm wrong about this, please correct me).

In changing the readme from markdown to reStructured text I've physically scaled the images and moved them to the github pages `_static` folder.

westurner commented:
http://docutils.sourceforge.net/docs/ref/rst/directives.html#image

There are :height:, :width:, and :scale: options; but IDK whether they work
w/ pypa/readme.

On Saturday, September 15, 2018, Peter Boin <notifications@github.com>
wrote:

> Hi all, I thought this was sorted months ago!
> But I totally left it hanging... my bad... so I'm finishing this asap.
>
> The issue is simple:
>
> reStructured text [when it comes to image properties] has very limited
> support by PyPI
>
> So simply scaling an image isn't possible (if I'm wrong about this, please
> correct me).
>
> In changing the readme from markdown to reStructured text I've physically
> scaled the images and moved them to the github pages _static folder.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dcowden/cadquery/pull/288#issuecomment-421544703>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AADGy_Qoj1Q3U65M647nt5bmTq7MI9yfks5ubMj-gaJpZM4VYBGw>
> .
>

westurner commented:
This is a much more useful README now; thanks

On Saturday, September 15, 2018, Wes Turner <wes.turner@gmail.com> wrote:

> http://docutils.sourceforge.net/docs/ref/rst/directives.html#image
>
> There are :height:, :width:, and :scale: options; but IDK whether they
> work w/ pypa/readme.
>
> On Saturday, September 15, 2018, Peter Boin <notifications@github.com>
> wrote:
>
>> Hi all, I thought this was sorted months ago!
>> But I totally left it hanging... my bad... so I'm finishing this asap.
>>
>> The issue is simple:
>>
>> reStructured text [when it comes to image properties] has very limited
>> support by PyPI
>>
>> So simply scaling an image isn't possible (if I'm wrong about this,
>> please correct me).
>>
>> In changing the readme from markdown to reStructured text I've physically
>> scaled the images and moved them to the github pages _static folder.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/dcowden/cadquery/pull/288#issuecomment-421544703>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AADGy_Qoj1Q3U65M647nt5bmTq7MI9yfks5ubMj-gaJpZM4VYBGw>
>> .
>>
>

fragmuffin commented:
@westurner 

I made a test PyPI package to test reStructured text rendering in [github](https://github.com/fragmuffin/pypi-test) vs [pypi test server](https://test.pypi.org/project/fragmuffin-pypitest)
That wasn't very successful; the rendering support for images pretty much stops at `src`, `alt`, and `href`. I couldn't get any images to scale in both, so I just did it physically.

westurner commented:
https://github.com/pypa/readme_renderer/pull/91 "Allow width, height, alt, and align attributes on img tags"

> Note: warehouse sets max-width: 100% for all img tags within the project description. This prevents users from being able to completely break the page layout if they set width to something ludicrous such as 100000%.

fragmuffin commented:
@westurner that doesn't match with what I'm seeing in the test package.
Any attempt to scale or centre images fails in both GitHub and PyPI :cry:

fragmuffin commented:
@westurner what do you think? is it time to merge?

westurner commented on README.rst:
```diff
@@ -0,0 +1,297 @@
+Thanks go to cadquery contributor hyOzd ( Altu Technology ) for the example!
+
+
+KiCad uses cadquery to build high quality models of electrictronic components. (`https://github.com/KiCad/packages3D <https://github.com/KiCad/packages3D>`_)
```
'electrictronic'

westurner commented on README.rst:
```diff
@@ -0,0 +1,297 @@
+
+CadQuery has several goals:
+
+* Build lD models with scripts that are as close as possible to how you'd describe the object to a human.
```
IDK what this is: "lD models"

westurner commented on README.rst:
```diff
@@ -0,0 +1,297 @@
+
+If you are familiar with jQuery, you will probably recognize several jQuery features that CadQuery uses:
+
+* A fluent api to create clean, easy to read code
```
api -> API

westurner commented on README.rst:
```diff
@@ -0,0 +1,297 @@
+
+CadQuery scripts have several key advantages over OpenSCAD:
+
+#. **The scripts use a standard programming language**, python, and thus can benefit from the associated infrastructure.
```
python -> Python

westurner commented:
Conda install instructions would be great, too:

```bash
conda install -c conda-forge freecad notebook cadquery?
```

fragmuffin commented:
@westurner good suggestions ;)
I've added the `conda` instruction in cdcea63 (which will publish the [`installation.html`](http://dcowden.github.io/cadquery/installation.html) page) ... I've simplified it to:

```
conda install -c conda-forge cadquery
```

Although I didn't test it...
@dcowden @jmwright @adam-urbanczyk : Is that the recommended way to install cadquery using anaconda?; or do you have to explicitly reference FreeCAD as well when installing?

adam-urbanczyk commented:
@fragmuffin it should be
```
conda install -c conda-forge cadquery
```
This works, but does not install all dependencies (e.g. FreeCad). So I have to conclude that the package is currently broken. Does anyone remember how was the conda-forge recipe prepared?

adam-urbanczyk commented:
Adding for reference: https://github.com/conda-forge/cadquery-feedstock/pull/2

fragmuffin commented:
@adam-urbanczyk excellent, thanks.
So I think this is ready to merge... @dcowden @jmwright what do you think?

jmwright commented:
@fragmuffin @adam-urbanczyk 

I have not used rst before. I'm assuming I'll have to learn its new markup format?

If the broken conda line isn't a showstopper, I think this is fine to merge.

fragmuffin commented:
@jmwright 

> I'm assuming I'll have to learn its new markup format?

Yes, it's a little different to markdown, but it has the same sentiment; the non-formatted text still gives a sense of hierarchy and format so it's easy to read.
I highly recommend learning it, especially when coding `sphinx` with Python, which `cadquery` does.

I often get markups mixed up, so I use an [online editor](http://rst.ninjs.org/) for basic verification of rst format.
However, as this PR has demonstrated, there are nuances with different renderers that can cause issues.
But I think we've accurately assessed that the benefits of `rst` outweigh the negatives, in this case.

I've added a check to TravisCI that should fail a bad edit to the `README.rst`

jmwright commented:
@dcowden Are you ok with this being merged?

dcowden commented:
@jmwright @fragmuffin definitely!

jmwright commented:
@fragmuffin @westurner @adam-urbanczyk Thanks for the work on this.

westurner commented:
Thanks!

On Wednesday, September 19, 2018, Jeremy Wright <notifications@github.com>
wrote:

> @fragmuffin <https://github.com/fragmuffin> @westurner
> <https://github.com/westurner> @adam-urbanczyk
> <https://github.com/adam-urbanczyk> Thanks for the work on this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dcowden/cadquery/pull/288#issuecomment-422898625>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AADGyyJh-NopDUH4BrV5bv3XrTdPjNj0ks5ucoTUgaJpZM4VYBGw>
> .
>

westurner commented:
This could be an https link:

http://dcowden.github.io/cadquery

On Wednesday, September 19, 2018, Wes Turner <wes.turner@gmail.com> wrote:

> Thanks!
>
> On Wednesday, September 19, 2018, Jeremy Wright <notifications@github.com>
> wrote:
>
>> @fragmuffin <https://github.com/fragmuffin> @westurner
>> <https://github.com/westurner> @adam-urbanczyk
>> <https://github.com/adam-urbanczyk> Thanks for the work on this.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/dcowden/cadquery/pull/288#issuecomment-422898625>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AADGyyJh-NopDUH4BrV5bv3XrTdPjNj0ks5ucoTUgaJpZM4VYBGw>
>> .
>>
>

westurner commented:
The images in
https://github.com/dcowden/cadquery/blob/master/README.rst#projects-using-cadquery
are 404ing?

http://dcowden.github.io/cadquery/_static/hyOzd-finished_thumb.jpg

...

A link to OpenCascade might be good:
https://en.wikipedia.org/wiki/Open_Cascade_Technology

On Wednesday, September 19, 2018, Wes Turner <wes.turner@gmail.com> wrote:

> This could be an https link:
>
> http://dcowden.github.io/cadquery
>
> On Wednesday, September 19, 2018, Wes Turner <wes.turner@gmail.com> wrote:
>
>> Thanks!
>>
>> On Wednesday, September 19, 2018, Jeremy Wright <notifications@github.com>
>> wrote:
>>
>>> @fragmuffin <https://github.com/fragmuffin> @westurner
>>> <https://github.com/westurner> @adam-urbanczyk
>>> <https://github.com/adam-urbanczyk> Thanks for the work on this.
>>>
>>> —
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/dcowden/cadquery/pull/288#issuecomment-422898625>,
>>> or mute the thread
>>> <https://github.com/notifications/unsubscribe-auth/AADGyyJh-NopDUH4BrV5bv3XrTdPjNj0ks5ucoTUgaJpZM4VYBGw>
>>> .
>>>
>>

jmwright commented:
@westurner The images show up for me.

westurner commented:
Must've taken a minute for GH pages to check out master and flush the cache.

This isn't displaying:
https://github.com/RustyVermeer/cqnb/raw/master/showcase.gif

The readme links to
https://github.com/RustyVermeer/cqnb/blob/master/showcase.gif

IDK if it's still necessary to append ``?raw=true``?
https://github.com/RustyVermeer/cqnb/blob/master/showcase.gif?raw=true

On Wednesday, September 19, 2018, Jeremy Wright <notifications@github.com>
wrote:

> @westurner <https://github.com/westurner> The images show up for me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dcowden/cadquery/pull/288#issuecomment-422902947>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AADGy1entHir1WrAEDLFmjyw2fD-uRKVks5ucofIgaJpZM4VYBGw>
> .
>

adam-urbanczyk commented on doc/installation.rst:
```diff
@@ -21,22 +24,37 @@ On Ubuntu, you can type::
+
+::
+
+    conda install -c conda-forge cadquery
```
Yes, but I am trying to convince the maintainer of the conda recipe to add it as a dep. Cf. https://github.com/conda-forge/cadquery-feedstock/pull/2

fragmuffin commented on README.rst:
```diff
@@ -0,0 +1,297 @@
+  * as a plugin for FreeCAD
+  * using the Docker Image to operate CadQuery as a CLI
+  * as a plugin running on a Jupyter Notebook server
+  * a standalone installation
```
Hmm, no... don't know how that happened.

fragmuffin replied:
fixed in #294

fragmuffin commented:
The `showcase.gif` link is still an issue.
I've made #294 to fix it.
